File: vault-sdks/ios/Sources/VaultAuth/VaultAuth.swift
import Foundation
import UIKit
import AuthenticationServices

// MARK: - VaultAuth

public class VaultAuth: NSObject, ObservableObject {
    
    // MARK: - Singleton
    
    public static let shared = VaultAuth()
    
    // MARK: - Properties
    
    private var vaultClient: VaultClient?
    private var tokenStorage: TokenStorage
    private var pushManager: PushNotificationManager
    
    private var apiKey: String?
    private var baseURL: String?
    private var tenantId: String?
    
    @Published public private(set) var currentUser: User?
    @Published public private(set) var currentOrganization: Organization?
    @Published public private(set) var organizations: [Organization] = []
    @Published public private(set) var isLoading = false
    
    public var isAuthenticated: Bool {
        return currentUser != nil && getSession()?.isValid == true
    }
    
    private var webAuthSession: ASWebAuthenticationSession?
    
    // MARK: - Initialization
    
    public init(tokenStorage: TokenStorage = KeychainStorage()) {
        self.tokenStorage = tokenStorage
        self.pushManager = PushNotificationManager()
        super.init()
        
        // Try to restore session
        Task {
            await restoreSession()
        }
    }
    
    // MARK: - Configuration
    
    public func configure(
        apiKey: String,
        baseURL: String,
        tenantId: String? = nil,
        tokenStorage: TokenStorage? = nil
    ) {
        self.apiKey = apiKey
        self.baseURL = baseURL
        self.tenantId = tenantId
        
        if let tokenStorage = tokenStorage {
            self.tokenStorage = tokenStorage
        }
        
        guard let url = URL(string: baseURL) else {
            print("Invalid base URL: \(baseURL)")
            return
        }
        
        self.vaultClient = VaultClient(
            baseURL: url,
            apiKey: apiKey,
            tenantId: tenantId
        )
        
        self.pushManager.configure(with: vaultClient!)
        
        // Update client with stored token if available
        if let session = getSession() {
            vaultClient?.setAccessToken(session.accessToken)
        }
        
        // Process any pending push notifications
        pushManager.processPendingNotifications()
    }
    
    // MARK: - Authentication Methods
    
    @discardableResult
    public func login(email: String, password: String) async throws -> User {
        guard let client = vaultClient else {
            throw VaultAuthError.notConfigured
        }
        
        isLoading = true
        defer { isLoading = false }
        
        let request = LoginRequest(
            email: email,
            password: password,
            tenantId: tenantId
        )
        
        do {
            let response: AuthResponse = try await client.post("/auth/login", body: request)
            try await handleAuthResponse(response)
            return response.user
        } catch VaultAuthError.mfaRequired {
            // MFA is required - the caller needs to handle this
            throw VaultAuthError.mfaRequired
        } catch {
            throw error
        }
    }
    
    @discardableResult
    public func loginWithBiometric() async throws -> User {
        guard BiometricAuth.shared.canUseBiometricLogin else {
            throw VaultAuthError.biometricNotAvailable
        }
        
        guard let credentials = try await BiometricAuth.shared.getBiometricCredentials() else {
            throw VaultAuthError.biometricFailed
        }
        
        return try await login(email: credentials.email, password: credentials.password)
    }
    
    @discardableResult
    public func signup(email: String, password: String, name: String? = nil) async throws -> User {
        guard let client = vaultClient else {
            throw VaultAuthError.notConfigured
        }
        
        isLoading = true
        defer { isLoading = false }
        
        let request = SignupRequest(
            email: email,
            password: password,
            name: name,
            tenantId: tenantId
        )
        
        let response: AuthResponse = try await client.post("/auth/signup", body: request)
        try await handleAuthResponse(response)
        return response.user
    }
    
    public func logout() async throws {
        guard let client = vaultClient else {
            throw VaultAuthError.notConfigured
        }
        
        do {
            try await client.post("/auth/logout")
        } catch {
            // Continue with local logout even if server logout fails
        }
        
        // Unregister push notifications
        await pushManager.unregisterDevice()
        
        // Clear local session
        try clearSession()
    }
    
    // MARK: - Token Verification
    
    @discardableResult
    public func verifyToken(_ token: String) async throws -> User {
        guard let client = vaultClient else {
            throw VaultAuthError.notConfigured
        }
        
        let request = TokenVerificationRequest(token: token)
        let user: User = try await client.post("/auth/verify", body: request)
        
        return user
    }
    
    // MARK: - Session Management
    
    public func refreshSession() async throws {
        guard let client = vaultClient else {
            throw VaultAuthError.notConfigured
        }
        
        guard let session = getSession() else {
            throw VaultAuthError.tokenNotFound
        }
        
        isLoading = true
        defer { isLoading = false }
        
        let request = TokenRefreshRequest(refreshToken: session.refreshToken)
        
        do {
            let response: TokenRefreshResponse = try await client.post("/auth/refresh", body: request)
            
            let newSession = Session(
                accessToken: response.accessToken,
                refreshToken: session.refreshToken,
                expiresAt: response.expiresAt,
                tokenType: response.tokenType
            )
            
            try saveSession(newSession)
            client.setAccessToken(newSession.accessToken)
            
            // Refresh user info
            let user: User = try await client.get("/auth/me")
            await MainActor.run {
                self.currentUser = user
            }
        } catch {
            // Clear session on refresh failure
            try? clearSession()
            throw VaultAuthError.tokenRefreshFailed
        }
    }
    
    @discardableResult
    private func restoreSession() async -> Bool {
        guard let client = vaultClient ?? createClientFromStoredConfig(),
              let session = getSession() else {
            return false
        }
        
        // Check if session is valid or needs refresh
        if session.isValid {
            client.setAccessToken(session.accessToken)
            await MainActor.run {
                self.vaultClient = client
            }
            
            // Fetch current user
            do {
                let user: User = try await client.get("/auth/me")
                await MainActor.run {
                    self.currentUser = user
                }
                return true
            } catch {
                return false
            }
        } else if session.isExpiringSoon || !session.isValid {
            // Try to refresh
            do {
                try await refreshSession()
                return true
            } catch {
                return false
            }
        }
        
        return false
    }
    
    // MARK: - OAuth
    
    @discardableResult
    public func loginWithOAuth(
        provider: OAuthProvider,
        from viewController: UIViewController
    ) async throws -> User {
        guard let client = vaultClient,
              let baseURL = baseURL else {
            throw VaultAuthError.notConfigured
        }
        
        let redirectUri = "vaultauth://oauth/callback"
        let state = UUID().uuidString
        
        let request = OAuthRequest(
            provider: provider.rawValue,
            redirectUri: redirectUri,
            state: state,
            tenantId: tenantId
        )
        
        // Get OAuth URL from server
        let response: OAuthInitResponse = try await client.post("/auth/oauth/init", body: request)
        
        guard let authURL = URL(string: response.authorizationUrl) else {
            throw VaultAuthError.oauthFailed("Invalid authorization URL")
        }
        
        return try await withCheckedThrowingContinuation { continuation in
            let session = ASWebAuthenticationSession(
                url: authURL,
                callbackURLScheme: "vaultauth"
            ) { callbackURL, error in
                if let error = error {
                    if (error as NSError).code == ASWebAuthenticationSessionError.canceledLogin.rawValue {
                        continuation.resume(throwing: VaultAuthError.oauthCancelled)
                    } else {
                        continuation.resume(throwing: VaultAuthError.oauthFailed(error.localizedDescription))
                    }
                    return
                }
                
                guard let callbackURL = callbackURL else {
                    continuation.resume(throwing: VaultAuthError.oauthFailed("No callback URL"))
                    return
                }
                
                // Parse callback URL
                guard let components = URLComponents(url: callbackURL, resolvingAgainstBaseURL: true),
                      let code = components.queryItems?.first(where: { $0.name == "code" })?.value,
                      let returnedState = components.queryItems?.first(where: { $0.name == "state" })?.value,
                      returnedState == state else {
                    continuation.resume(throwing: VaultAuthError.oauthFailed("Invalid callback"))
                    return
                }
                
                // Exchange code for tokens
                Task {
                    do {
                        let callbackRequest = OAuthCallbackRequest(
                            code: code,
                            state: state,
                            tenantId: self.tenantId
                        )
                        
                        let authResponse: AuthResponse = try await client.post("/auth/oauth/callback", body: callbackRequest)
                        try await self.handleAuthResponse(authResponse)
                        continuation.resume(returning: authResponse.user)
                    } catch {
                        continuation.resume(throwing: error)
                    }
                }
            }
            
            session.presentationContextProvider = self
            session.prefersEphemeralWebBrowserSession = false
            
            self.webAuthSession = session
            session.start()
        }
    }
    
    // MARK: - MFA
    
    public func enableMFA(method: MFAMethod) async throws -> MFAEnableResponse {
        guard let client = vaultClient else {
            throw VaultAuthError.notConfigured
        }
        
        let request = MFAEnableRequest(method: method.rawValue)
        let response: MFAEnableResponse = try await client.post("/auth/mfa/enable", body: request)
        return response
    }
    
    @discardableResult
    public func verifyMFA(code: String, method: MFAMethod? = nil) async throws -> User {
        guard let client = vaultClient else {
            throw VaultAuthError.notConfigured
        }
        
        let request = MFAVerifyRequest(code: code, method: method?.rawValue)
        let response: AuthResponse = try await client.post("/auth/mfa/verify", body: request)
        try await handleAuthResponse(response)
        return response.user
    }
    
    public func disableMFA(method: MFAMethod, code: String) async throws {
        guard let client = vaultClient else {
            throw VaultAuthError.notConfigured
        }
        
        let request = MFADisableRequest(code: code, method: method.rawValue)
        try await client.post("/auth/mfa/disable", body: request)
    }
    
    // MARK: - Organizations
    
    public func getOrganizations() async throws -> [Organization] {
        guard let client = vaultClient else {
            throw VaultAuthError.notConfigured
        }
        
        let response: OrganizationListResponse = try await client.get("/auth/organizations")
        
        await MainActor.run {
            self.organizations = response.organizations
        }
        
        return response.organizations
    }
    
    @discardableResult
    public func switchOrganization(_ organizationId: String) async throws -> Organization {
        guard let client = vaultClient else {
            throw VaultAuthError.notConfigured
        }
        
        let request = OrganizationSwitchRequest(organizationId: organizationId)
        let organization: Organization = try await client.post("/auth/organizations/switch", body: request)
        
        await MainActor.run {
            self.currentOrganization = organization
        }
        
        // Store current organization
        try? tokenStorage.saveToken(organizationId, forKey: TokenStorageKeys.organizationId)
        
        return organization
    }
    
    // MARK: - Push Notifications
    
    public func registerForPushNotifications(deviceToken: Data) {
        pushManager.registerDevice(token: deviceToken)
    }
    
    public func handlePushNotification(_ notification: [AnyHashable: Any]) {
        pushManager.handleNotification(notification)
    }
    
    public var pushNotificationHandler: PushNotificationHandler? {
        guard let client = vaultClient else { return nil }
        return PushNotificationHandler(vaultClient: client)
    }
    
    // MARK: - Private Helpers
    
    private func handleAuthResponse(_ response: AuthResponse) async throws {
        // Save session
        try saveSession(response.session)
        
        // Update client with new token
        vaultClient?.setAccessToken(response.session.accessToken)
        
        // Update published properties on main thread
        await MainActor.run {
            self.currentUser = response.user
        }
        
        // Fetch organizations
        do {
            let orgs = try await getOrganizations()
            if let firstOrg = orgs.first, currentOrganization == nil {
                _ = try? await switchOrganization(firstOrg.id)
            }
        } catch {
            // Non-critical error
        }
    }
    
    private func saveSession(_ session: Session) throws {
        if let keychainStorage = tokenStorage as? KeychainStorage {
            try keychainStorage.saveSession(session)
        } else {
            try tokenStorage.saveToken(session.accessToken, forKey: TokenStorageKeys.accessToken)
            try tokenStorage.saveToken(session.refreshToken, forKey: TokenStorageKeys.refreshToken)
            try tokenStorage.saveToken(String(session.expiresAt.timeIntervalSince1970), forKey: TokenStorageKeys.sessionExpiresAt)
        }
        
        vaultClient?.setAccessToken(session.accessToken)
    }
    
    private func getSession() -> Session? {
        if let keychainStorage = tokenStorage as? KeychainStorage {
            return keychainStorage.getSession()
        }
        
        guard let accessToken = tokenStorage.getToken(forKey: TokenStorageKeys.accessToken),
              let refreshToken = tokenStorage.getToken(forKey: TokenStorageKeys.refreshToken),
              let expiresAtString = tokenStorage.getToken(forKey: TokenStorageKeys.sessionExpiresAt),
              let expiresAt = TimeInterval(expiresAtString) else {
            return nil
        }
        
        return Session(
            accessToken: accessToken,
            refreshToken: refreshToken,
            expiresAt: Date(timeIntervalSince1970: expiresAt)
        )
    }
    
    private func clearSession() throws {
        if let keychainStorage = tokenStorage as? KeychainStorage {
            try keychainStorage.deleteSession()
        } else {
            try tokenStorage.deleteToken(forKey: TokenStorageKeys.accessToken)
            try tokenStorage.deleteToken(forKey: TokenStorageKeys.refreshToken)
            try tokenStorage.deleteToken(forKey: TokenStorageKeys.sessionExpiresAt)
        }
        
        try tokenStorage.deleteToken(forKey: TokenStorageKeys.organizationId)
        
        await MainActor.run {
            self.currentUser = nil
            self.currentOrganization = nil
            self.organizations = []
        }
        
        vaultClient?.setAccessToken(nil)
    }
    
    private func createClientFromStoredConfig() -> VaultClient? {
        // This would need to store/retrieve config from UserDefaults or similar
        // For now, return nil
        return nil
    }
}

// MARK: - ASWebAuthenticationPresentationContextProviding

extension VaultAuth: ASWebAuthenticationPresentationContextProviding {
    public func presentationAnchor(for session: ASWebAuthenticationSession) -> ASPresentationAnchor {
        return UIApplication.shared.windows.first { $0.isKeyWindow } ?? UIWindow()
    }
}

// MARK: - OAuthInitResponse

struct OAuthInitResponse: Decodable {
    let authorizationUrl: String
}

File: packages/sdks/admin-sdks/android/vaultauth/src/main/java/com.fantasticauth/VaultAuth.kt
package com.vault

import android.app.Activity
import android.content.Context
import android.content.Intent
import androidx.fragment.app.FragmentActivity
import com.vault.biometric.BiometricAuth
import com.vault.biometric.BiometricType
import com.vault.models.*
import com.vault.storage.*
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.withContext
import com.google.gson.Gson
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

/**
 * Main Vault Authentication SDK class
 * 
 * Usage:
 * ```kotlin
 * // Initialize in Application.onCreate()
 * VaultAuth.getInstance().configure(
 *     context = applicationContext,
 *     apiKey = "your_api_key",
 *     baseUrl = "https://vault.example.com",
 *     tenantId = "your_tenant_id"
 * )
 * 
 * // Login
 * val user = VaultAuth.getInstance().login("user@example.com", "password")
 * ```
 */
class VaultAuth private constructor() {

    private var vaultClient: VaultClient? = null
    private var tokenStorage: TokenStorage? = null
    private var config: VaultConfig? = null
    private val gson = Gson()
    private val mutex = Mutex()

    companion object {
        @Volatile
        private var instance: VaultAuth? = null

        /**
         * Get the singleton instance of VaultAuth
         */
        @JvmStatic
        fun getInstance(): VaultAuth {
            return instance ?: synchronized(this) {
                instance ?: VaultAuth().also { instance = it }
            }
        }

        /**
         * Clear the singleton instance (useful for testing)
         */
        @JvmStatic
        fun clearInstance() {
            instance = null
        }
    }

    /**
     * Configure the VaultAuth SDK
     * 
     * @param context Application context
     * @param apiKey Your Vault API key
     * @param baseUrl The base URL of your Vault server
     * @param tenantId Optional tenant ID for multi-tenant setups
     * @param debug Enable debug logging
     * @param storage Custom token storage implementation (optional)
     */
    @JvmOverloads
    fun configure(
        context: Context,
        apiKey: String,
        baseUrl: String,
        tenantId: String? = null,
        debug: Boolean = false,
        storage: TokenStorage? = null
    ) {
        if (vaultClient != null) {
            throw VaultAuthException.InvalidConfiguration("VaultAuth is already configured")
        }

        val normalizedBaseUrl = if (baseUrl.endsWith("/")) baseUrl else "$baseUrl/"
        
        config = VaultConfig(
            apiKey = apiKey,
            baseUrl = normalizedBaseUrl,
            tenantId = tenantId,
            debug = debug
        )

        tokenStorage = storage ?: KeystoreStorage(context.applicationContext)
        vaultClient = VaultClient(tokenStorage!!, config!!)
        
        // Try to restore previous session
        vaultClient?.restoreSession()
    }

    /**
     * Check if the SDK has been configured
     */
    fun isConfigured(): Boolean = vaultClient != null

    private fun getClient(): VaultClient {
        return vaultClient ?: throw VaultAuthException.NotConfigured()
    }

    private fun getStorage(): TokenStorage {
        return tokenStorage ?: throw VaultAuthException.NotConfigured()
    }

    // ============================================================================
    // Authentication State
    // ============================================================================

    /**
     * Current authentication state flow
     */
    val authState: StateFlow<AuthState>?
        get() = vaultClient?.authState

    /**
     * Current authenticated user
     */
    val currentUser: User?
        get() = vaultClient?.currentUser?.value

    /**
     * Current session
     */
    val currentSession: Session?
        get() = vaultClient?.currentSession?.value

    /**
     * Current organization
     */
    val currentOrganization: Organization?
        get() = vaultClient?.currentOrganization?.value

    /**
     * Check if user is currently authenticated
     */
    val isAuthenticated: Boolean
        get() = vaultClient?.isAuthenticated() == true

    // ============================================================================
    // Authentication Methods
    // ============================================================================

    /**
     * Login with email and password
     * 
     * @param email User's email
     * @param password User's password
     * @return The authenticated User
     * @throws VaultAuthException if authentication fails
     */
    suspend fun login(email: String, password: String): User = withContext(Dispatchers.IO) {
        val client = getClient()
        val tenantId = config?.tenantId

        val response = client.api.login(LoginRequest(email, password, tenantId))

        if (response.isSuccessful) {
            val authResponse = response.body()
                ?: throw VaultAuthException.DecodingError(Exception("Empty response body"))
            
            client.setAuthenticated(authResponse.user, authResponse.session)
            authResponse.user
        } else {
            val errorBody = response.errorBody()?.string()
            val apiError = errorBody?.let { 
                try {
                    gson.fromJson(it, ApiErrorResponse::class.java)
                } catch (e: Exception) {
                    ApiErrorResponse("unknown", errorBody)
                }
            } ?: ApiErrorResponse("unknown", "Unknown error")
            
            throw apiError.toException(response.code())
        }
    }

    /**
     * Login with biometric authentication
     * Requires that biometric login has been previously enabled
     * 
     * @param activity The activity for showing biometric prompt
     * @return The authenticated User
     * @throws VaultAuthException if authentication fails
     */
    suspend fun loginWithBiometric(activity: FragmentActivity): User = withContext(Dispatchers.Main) {
        val biometricAuth = BiometricAuth(activity)
        
        if (!biometricAuth.isAvailable) {
            throw VaultAuthException.BiometricNotAvailable()
        }

        val storage = getStorage()
        
        // First authenticate with biometric
        biometricAuth.authenticate(
            title = "Log In",
            subtitle = "Use your biometric to log in",
            negativeButtonText = "Use Password"
        )

        // Get the biometric-protected token
        val biometricToken = storage.getToken(TokenStorage.KEY_BIOMETRIC_TOKEN)
            ?: throw VaultAuthException.TokenNotFound()

        // Verify the token with the server
        verifyToken(biometricToken)
    }

    /**
     * Sign up a new user
     * 
     * @param email User's email
     * @param password User's password
     * @param name User's full name (optional)
     * @return The created User
     * @throws VaultAuthException if signup fails
     */
    suspend fun signup(email: String, password: String, name: String? = null): User = withContext(Dispatchers.IO) {
        val client = getClient()
        val tenantId = config?.tenantId

        val response = client.api.signup(SignupRequest(email, password, name, tenantId))

        if (response.isSuccessful) {
            val authResponse = response.body()
                ?: throw VaultAuthException.DecodingError(Exception("Empty response body"))
            
            client.setAuthenticated(authResponse.user, authResponse.session)
            authResponse.user
        } else {
            val errorBody = response.errorBody()?.string()
            val apiError = errorBody?.let { 
                try {
                    gson.fromJson(it, ApiErrorResponse::class.java)
                } catch (e: Exception) {
                    ApiErrorResponse("unknown", errorBody)
                }
            } ?: ApiErrorResponse("unknown", "Unknown error")
            
            throw apiError.toException(response.code())
        }
    }

    /**
     * Logout the current user
     */
    suspend fun logout() = withContext(Dispatchers.IO) {
        val client = getClient()
        
        try {
            client.api.logout()
        } catch (e: Exception) {
            // Continue with local logout even if server logout fails
        } finally {
            client.clearAuthentication()
        }
    }

    /**
     * Verify a token with the server
     * 
     * @param token The token to verify
     * @return The user associated with the token
     */
    suspend fun verifyToken(token: String): User = withContext(Dispatchers.IO) {
        val client = getClient()

        val response = client.api.verifyToken(TokenVerificationRequest(token))

        if (response.isSuccessful) {
            response.body() ?: throw VaultAuthException.DecodingError(Exception("Empty response body"))
        } else {
            val errorBody = response.errorBody()?.string()
            val apiError = errorBody?.let { 
                try {
                    gson.fromJson(it, ApiErrorResponse::class.java)
                } catch (e: Exception) {
                    ApiErrorResponse("unknown", errorBody)
                }
            } ?: ApiErrorResponse("unknown", "Unknown error")
            
            throw apiError.toException(response.code())
        }
    }

    /**
     * Refresh the current session
     */
    suspend fun refreshSession() = withContext(Dispatchers.IO) {
        val client = getClient()
        val storage = getStorage()
        
        val refreshToken = storage.getToken(TokenStorage.KEY_REFRESH_TOKEN)
            ?: throw VaultAuthException.TokenNotFound()

        val response = client.api.refreshToken(TokenRefreshRequest(refreshToken))

        if (response.isSuccessful) {
            val refreshResponse = response.body()
                ?: throw VaultAuthException.DecodingError(Exception("Empty response body"))
            
            val currentUser = client.currentUser.value
                ?: throw VaultAuthException.SessionExpired()
            
            val newSession = Session(
                id = currentSession?.id ?: "",
                accessToken = refreshResponse.accessToken,
                refreshToken = refreshResponse.refreshToken ?: refreshToken,
                expiresAt = refreshResponse.expiresAt,
                user = currentUser
            )
            
            client.updateSession(newSession)
        } else {
            client.clearAuthentication()
            throw VaultAuthException.TokenRefreshFailed()
        }
    }

    // ============================================================================
    // OAuth
    // ============================================================================

    /**
     * Login with OAuth provider
     * 
     * @param provider The OAuth provider (GOOGLE, GITHUB, MICROSOFT, etc.)
     * @param activity The activity for OAuth flow
     * @return The authenticated User
     */
    suspend fun loginWithOAuth(provider: OAuthProvider, activity: Activity): User {
        // This would typically launch a custom tab or WebView for OAuth
        // Implementation depends on the specific OAuth flow
        throw VaultAuthException.OAuthFailed("OAuth not implemented. Use OAuth libraries like AppAuth.")
    }

    /**
     * Handle OAuth callback
     * Call this from your OAuth callback activity
     * 
     * @param intent The intent containing the OAuth callback
     * @return The authenticated User
     */
    suspend fun handleOAuthCallback(intent: Intent): User {
        // Parse the OAuth callback and exchange code for tokens
        throw VaultAuthException.OAuthFailed("OAuth not implemented. Use OAuth libraries like AppAuth.")
    }

    // ============================================================================
    // MFA
    // ============================================================================

    /**
     * Enable MFA for the current user
     * 
     * @param method The MFA method to enable
     * @return MFA setup information (including QR code for TOTP)
     */
    suspend fun enableMfa(method: MfaMethod): MfaSetup = withContext(Dispatchers.IO) {
        val client = getClient()

        if (method == MfaMethod.TOTP) {
            val response = client.api.setupMfa(TotpSetupRequest())
            
            if (response.isSuccessful) {
                response.body() ?: throw VaultAuthException.DecodingError(Exception("Empty response body"))
            } else {
                val errorBody = response.errorBody()?.string()
                val apiError = errorBody?.let { 
                    try {
                        gson.fromJson(it, ApiErrorResponse::class.java)
                    } catch (e: Exception) {
                        ApiErrorResponse("unknown", errorBody)
                    }
                } ?: ApiErrorResponse("unknown", "Unknown error")
                
                throw apiError.toException(response.code())
            }
        } else {
            val response = client.api.enableMfa(EnableMfaRequest(method))
            
            if (!response.isSuccessful) {
                val errorBody = response.errorBody()?.string()
                throw VaultAuthException.MfaSetupFailed(errorBody ?: "Failed to enable MFA")
            }
            
            MfaSetup("", "", emptyList())
        }
    }

    /**
     * Verify MFA code during login
     * 
     * @param code The MFA code
     * @param method The MFA method
     */
    suspend fun verifyMfa(code: String, method: MfaMethod = MfaMethod.TOTP): Unit = withContext(Dispatchers.IO) {
        val client = getClient()

        val response = client.api.verifyMfa(MfaVerifyRequest(code, method))

        if (response.isSuccessful) {
            val authResponse = response.body()
                ?: throw VaultAuthException.DecodingError(Exception("Empty response body"))
            
            client.setAuthenticated(authResponse.user, authResponse.session)
        } else {
            val errorBody = response.errorBody()?.string()
            val apiError = errorBody?.let { 
                try {
                    gson.fromJson(it, ApiErrorResponse::class.java)
                } catch (e: Exception) {
                    ApiErrorResponse("unknown", errorBody)
                }
            } ?: ApiErrorResponse("unknown", "Unknown error")
            
            throw apiError.toException(response.code())
        }
    }

    /**
     * Disable MFA for the current user
     * 
     * @param method The MFA method to disable
     */
    suspend fun disableMfa(method: MfaMethod) = withContext(Dispatchers.IO) {
        val client = getClient()

        val response = client.api.disableMfa(EnableMfaRequest(method))

        if (!response.isSuccessful) {
            val errorBody = response.errorBody()?.string()
            throw VaultAuthException.MfaSetupFailed(errorBody ?: "Failed to disable MFA")
        }
    }

    // ============================================================================
    // Organizations
    // ============================================================================

    /**
     * Get all organizations for the current user
     * 
     * @return List of organizations
     */
    suspend fun getOrganizations(): List<Organization> = withContext(Dispatchers.IO) {
        val client = getClient()

        val response = client.api.getOrganizations()

        if (response.isSuccessful) {
            val orgResponse = response.body()
            orgResponse?.organizations ?: emptyList()
        } else {
            val errorBody = response.errorBody()?.string()
            val apiError = errorBody?.let { 
                try {
                    gson.fromJson(it, ApiErrorResponse::class.java)
                } catch (e: Exception) {
                    ApiErrorResponse("unknown", errorBody)
                }
            } ?: ApiErrorResponse("unknown", "Unknown error")
            
            throw apiError.toException(response.code())
        }
    }

    /**
     * Switch to a different organization
     * 
     * @param organizationId The ID of the organization to switch to
     */
    suspend fun switchOrganization(organizationId: String) = withContext(Dispatchers.IO) {
        val client = getClient()

        val response = client.api.switchOrganization(SwitchOrganizationRequest(organizationId))

        if (response.isSuccessful) {
            val switchResponse = response.body()
                ?: throw VaultAuthException.DecodingError(Exception("Empty response body"))
            
            client.setCurrentOrganization(switchResponse.organization)
            client.updateSession(switchResponse.session)
        } else {
            throw VaultAuthException.OrganizationSwitchFailed()
        }
    }

    // ============================================================================
    // User Management
    // ============================================================================

    /**
     * Update the current user
     * 
     * @param request The update request
     * @return The updated User
     */
    suspend fun updateUser(request: UserUpdateRequest): User = withContext(Dispatchers.IO) {
        val client = getClient()

        val response = client.api.updateUser(request)

        if (response.isSuccessful) {
            response.body() ?: throw VaultAuthException.DecodingError(Exception("Empty response body"))
        } else {
            val errorBody = response.errorBody()?.string()
            val apiError = errorBody?.let { 
                try {
                    gson.fromJson(it, ApiErrorResponse::class.java)
                } catch (e: Exception) {
                    ApiErrorResponse("unknown", errorBody)
                }
            } ?: ApiErrorResponse("unknown", "Unknown error")
            
            throw apiError.toException(response.code())
        }
    }

    /**
     * Change the current user's password
     * 
     * @param currentPassword The current password
     * @param newPassword The new password
     */
    suspend fun changePassword(currentPassword: String, newPassword: String) = withContext(Dispatchers.IO) {
        val client = getClient()

        val response = client.api.changePassword(PasswordChangeRequest(currentPassword, newPassword))

        if (!response.isSuccessful) {
            val errorBody = response.errorBody()?.string()
            val apiError = errorBody?.let { 
                try {
                    gson.fromJson(it, ApiErrorResponse::class.java)
                } catch (e: Exception) {
                    ApiErrorResponse("unknown", errorBody)
                }
            } ?: ApiErrorResponse("unknown", "Unknown error")
            
            throw apiError.toException(response.code())
        }
    }

    // ============================================================================
    // Session Management
    // ============================================================================

    /**
     * Get all active sessions for the current user
     * 
     * @return List of session information
     */
    suspend fun getSessions(): List<SessionInfo> = withContext(Dispatchers.IO) {
        val client = getClient()

        val response = client.api.getSessions()

        if (response.isSuccessful) {
            response.body() ?: emptyList()
        } else {
            emptyList()
        }
    }

    /**
     * Revoke a specific session
     * 
     * @param sessionId The ID of the session to revoke
     */
    suspend fun revokeSession(sessionId: String) = withContext(Dispatchers.IO) {
        val client = getClient()
        client.api.revokeSession(sessionId)
    }

    /**
     * Revoke all other sessions except the current one
     */
    suspend fun revokeAllOtherSessions() = withContext(Dispatchers.IO) {
        val client = getClient()
        val sessions = getSessions()
        
        val currentSessionId = currentSession?.id
        
        sessions.forEach { session ->
            if (!session.isCurrent && session.id != currentSessionId) {
                try {
                    client.api.revokeSession(session.id)
                } catch (e: Exception) {
                    // Continue revoking other sessions
                }
            }
        }
    }

    // ============================================================================
    // Biometric
    // ============================================================================

    /**
     * Enable biometric login for the current user
     * 
     * @param activity The activity for showing biometric prompt
     */
    suspend fun enableBiometricLogin(activity: FragmentActivity) = withContext(Dispatchers.Main) {
        val biometricAuth = BiometricAuth(activity)
        
        if (!biometricAuth.isAvailable) {
            throw VaultAuthException.BiometricNotAvailable()
        }

        // Authenticate with biometric
        biometricAuth.authenticate(
            title = "Enable Biometric Login",
            subtitle = "Verify your identity to enable biometric login",
            description = "You'll be able to log in using your biometric credentials in the future."
        )

        // Store a special token for biometric login
        val session = currentSession ?: throw VaultAuthException.SessionExpired()
        val storage = getStorage()
        
        // Store the refresh token with biometric protection
        storage.saveToken(TokenStorage.KEY_BIOMETRIC_TOKEN, session.refreshToken)
    }

    /**
     * Disable biometric login
     */
    fun disableBiometricLogin() {
        val storage = getStorage()
        storage.deleteToken(TokenStorage.KEY_BIOMETRIC_TOKEN)
    }

    /**
     * Check if biometric login is enabled
     */
    fun isBiometricLoginEnabled(): Boolean {
        return getStorage().hasToken(TokenStorage.KEY_BIOMETRIC_TOKEN)
    }

    /**
     * Check if biometric authentication is available on this device
     */
    fun isBiometricAvailable(context: Context): Boolean {
        return BiometricAuth.isAvailable(context)
    }

    /**
     * Get the type of biometric available
     */
    fun getBiometricType(context: Context): BiometricType {
        return BiometricAuth.getBiometricType(context)
    }

    // ============================================================================
    // Push Notifications
    // ============================================================================

    /**
     * Register for push notifications
     * 
     * @param token The FCM token
     */
    suspend fun registerForPushNotifications(token: String) = withContext(Dispatchers.IO) {
        val client = getClient()
        
        try {
            client.api.registerPushToken(PushTokenRequest(token))
        } catch (e: Exception) {
            // Log but don't throw - push registration is not critical
        }
    }

    /**
     * Approve an MFA push request
     * 
     * @param requestId The MFA request ID
     */
    suspend fun approveMfaRequest(requestId: String) = withContext(Dispatchers.IO) {
        val client = getClient()
        client.api.approveMfaPush(MfaPushRequest(requestId))
    }

    /**
     * Deny an MFA push request
     * 
     * @param requestId The MFA request ID
     */
    suspend fun denyMfaRequest(requestId: String) = withContext(Dispatchers.IO) {
        val client = getClient()
        client.api.denyMfaPush(MfaPushRequest(requestId))
    }
}
